crates/binstalk-downloader/src/gh_api_client/request.rs

```diff
@@ -102,27 +131,133 @@ pub(super) async fn fetch_release_artifacts(
     let status = response.status();
     let headers = response.headers();
 
-    if status == remote::StatusCode::FORBIDDEN
-        && headers
-            .get("x-ratelimit-remaining")
-            .map(|val| val == "0")
-            .unwrap_or(false)
-    {
-        return Ok(FetchReleaseRet::ReachedRateLimit {
-            retry_after: headers.get("x-ratelimit-reset").and_then(|value| {
-                let secs = value.to_str().ok()?.parse().ok()?;
-                Some(Duration::from_secs(secs))
-            }),
-        });
+    if let Some(ret) = check_for_status(status, headers) {
+        return Ok(ret);
     }
 
-    if status == remote::StatusCode::UNAUTHORIZED {
-        return Ok(FetchReleaseRet::Unauthorized);
+    Ok(FetchReleaseRet::Artifacts(response.json().await?))
+}
+
+#[derive(Deserialize)]
+struct GraphQLResponse {
+    data: GraphQLData,
+}
+
+#[derive(Deserialize)]
+struct GraphQLData {
+    repository: Option<GraphQLRepo>,
+}
+
+#[derive(Deserialize)]
+struct GraphQLRepo {
+    release: Option<GraphQLRelease>,
+}
+
+#[derive(Deserialize)]
+struct GraphQLRelease {
+    #[serde(rename = "releaseAssets")]
+    assets: GraphQLReleaseAssets,
+}
+
+#[derive(Deserialize)]
+struct GraphQLReleaseAssets {
+    nodes: Vec<Artifact>,
+    #[serde(rename = "pageInfo")]
+    page_info: GraphQLPageInfo,
+}
+
+#[derive(Deserialize)]
+struct GraphQLPageInfo {
+    #[serde(rename = "endCursor")]
+    end_cursor: CompactString,
+    #[serde(rename = "hasNextPage")]
+    has_next_page: bool,
+}
+
+enum FilterCondition {
+    Init,
+    After(CompactString),
+}
+
+impl fmt::Display for FilterCondition {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            // GitHub imposes a limit of 100 for the value passed to param "first"
+            FilterCondition::Init => f.write_str("first:100"),
+            FilterCondition::After(end_cursor) => write!(f, "after:\\\"{end_cursor}\\\""),
+        }
     }
+}
 
-    if status == remote::StatusCode::NOT_FOUND {
-        return Ok(FetchReleaseRet::ReleaseNotFound);
+async fn fetch_release_artifacts_graphql_api(
+    client: &remote::Client,
+    GhRelease { owner, repo, tag }: &GhRelease,
```

Comment:
I think with graphql we could build the query in such a way that we could query for _every_ desired package at once, which would be a marked improvement, but can reserve that for a future go.

Reply:
That's definitely possible as you have demonstrated in https://github.com/cargo-bins/cargo-binstall/pull/832#discussion_r1117895247 but that would require refactoring the way fetchers work, so yeah, let's save for future.

Reply:
Actually it's a lot of work, plus it might not actually worth the complexity added.
And in case we need to fallback to restful API or plain GET for each url, it's just the same as before.